drivers/rtc/rtc_sam0.c

```diff
@@ -0,0 +1,649 @@
+
+	return 0;
+}
+
+static int rtc_sam0_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				   struct rtc_time *timeptr)
+{
+	const struct rtc_sam0_config *cfg = dev->config;
+	struct rtc_sam0_data *data = dev->data;
+	RtcMode2 *regs = &cfg->regs->MODE2;
+	RTC_MODE2_ALARM_Type datetime;
+	uint32_t alarm_msk;
+
+	if (BIT(id) > RTC_MODE2_INTFLAG_ALARM_Msk) {
+		return -EINVAL;
+	}
+
+	if ((mask == NULL) || (timeptr == NULL)) {
+		return -EINVAL;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	rtc_sam0_sync(cfg->regs);
+
+	datetime = regs->Mode2Alarm[id].ALARM;
+	alarm_msk = regs->Mode2Alarm[id].MASK.reg;
+
+	LOG_INF("G: datetime: %d, mask: %d", datetime.reg, alarm_msk);
+
+	k_spin_unlock(&data->lock, key);
+
+	*mask = rtc_sam0_mask_from_alarm_msk(alarm_msk);
+
+	rtc_sam0_tm_from_datetime(timeptr, *mask, datetime);
+
+	return 0;
+}
+
+static int rtc_sam0_alarm_is_pending(const struct device *dev, uint16_t id)
+{
+	const struct rtc_sam0_config *cfg = dev->config;
+	struct rtc_sam0_data *data = dev->data;
+	RtcMode2 *regs = &cfg->regs->MODE2;
+
+	if (BIT(id) > RTC_MODE2_INTFLAG_ALARM_Msk) {
+		return -EINVAL;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	if ((regs->INTFLAG.reg & RTC_MODE2_INTFLAG_ALARM(BIT(id))) == 0) {
+		k_spin_unlock(&data->lock, key);
+
+		return 0;
+	}
+
+	regs->INTFLAG.reg = RTC_MODE2_INTFLAG_ALARM(BIT(id));
+
+	k_spin_unlock(&data->lock, key);
+
+	return 1;
+}
+
+static int rtc_sam0_alarm_set_callback(const struct device *dev, uint16_t id,
+				       rtc_alarm_callback callback, void *user_data)
+{
+	const struct rtc_sam0_config *cfg = dev->config;
+	struct rtc_sam0_data *data = dev->data;
+	RtcMode2 *regs = &cfg->regs->MODE2;
+
+	if (BIT(id) > RTC_MODE2_INTFLAG_ALARM_Msk) {
+		return -EINVAL;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	data->alarms[id].cb = callback;
+	data->alarms[id].cb_data = user_data;
+
+	if (callback) {
+		regs->INTENSET.reg = RTC_MODE2_INTENSET_ALARM(BIT(id));
+	} else {
+		regs->INTENCLR.reg = RTC_MODE2_INTENCLR_ALARM(BIT(id));
+	}
+
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+static void rtc_sam0_isr(const struct device *dev)
+{
+	const struct rtc_sam0_config *cfg = dev->config;
+	struct rtc_sam0_data *data = dev->data;
+	RtcMode2 *regs = &cfg->regs->MODE2;
+	uint32_t int_flags = regs->INTFLAG.reg;
+
+	for (int i = 0; i < cfg->alarms_count; ++i) {
+		if (int_flags & RTC_MODE2_INTFLAG_ALARM(BIT(i))) {
+			if (data->alarms[i].cb != NULL) {
+				data->alarms[i].cb(dev, i, data->alarms[i].cb_data);
+			}
+		}
+	}
+
+	regs->INTFLAG.reg |= int_flags;
+}
+
+#endif /* CONFIG_RTC_ALARM */
+
+#ifdef CONFIG_RTC_CALIBRATION
+static int rtc_sam0_set_calibration(const struct device *dev, int32_t calibration)
+{
+	const struct rtc_sam0_config *cfg = dev->config;
+	RtcMode2 *regs = &cfg->regs->MODE2;
+	int32_t correction = (int32_t)((float)cfg->cal_constant
+			   * (float)calibration
+			   * (float)1e-9)
+			   ;
```

Comment:
This does not look correct, using floats at all for integer calculations is not the most efficient way to do it :)

Is there supposed to be a division somewhere in here? Something like:
```
/* Correction in ppm = (FREQCORR.VALUE / 4096 * 240) * 1000000 */
=>
int32_t correction = calibration / (1000000000 / cfg->cal_constant);
=>
int32_t correction = calibration / 1017; /* a tiny loss of precision compared to float, waaay faster */
```

Reply:
Hi,

I'm waiting test feedback. As soon I have I'll update the PR.

Reply:
Done!